exp/addons/internal/controllers/clusterresourceset_helpers.go

```diff
@@ -55,56 +93,30 @@ func isJSONList(data []byte) (bool, error) {
 	return bytes.HasPrefix(trim, jsonListPrefix), nil
 }
 
-func apply(ctx context.Context, c client.Client, data []byte) error {
-	isJSONList, err := isJSONList(data)
-	if err != nil {
-		return err
-	}
-	objs := []unstructured.Unstructured{}
-	// If it is a json list, convert each list element to an unstructured object.
-	if isJSONList {
-		var results []map[string]interface{}
-		// Unmarshal the JSON to the interface.
-		if err = json.Unmarshal(data, &results); err == nil {
-			for i := range results {
-				var u unstructured.Unstructured
-				u.SetUnstructuredContent(results[i])
-				objs = append(objs, u)
-			}
-		}
-	} else {
-		// If it is not a json list, data is either json or yaml format.
-		objs, err = utilyaml.ToUnstructured(data)
-		if err != nil {
-			return errors.Wrapf(err, "failed converting data to unstructured objects")
-		}
-	}
-
+// apply reconciles all objects defined by a resource from a ClusterResourceSet
+// delegating on the reconcileScope for the applying strategy.
+func apply(ctx context.Context, c client.Client, scope resourceReconcileScope) error {
 	errList := []error{}
-	sortedObjs := utilresource.SortForCreate(objs)
-	for i := range sortedObjs {
-		if err := applyUnstructured(ctx, c, &sortedObjs[i]); err != nil {
+	objs := scope.objs()
+	for i := range objs {
+		if err := scope.apply(ctx, c, &objs[i]); err != nil {
 			errList = append(errList, err)
 		}
 	}
 	return kerrors.NewAggregate(errList)
 }
 
-func applyUnstructured(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
-	// Create the object on the API server.
-	// TODO: Errors are only logged. If needed, exponential backoff or requeuing could be used here for remedying connection glitches etc.
+func createUnstructured(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
 	if err := c.Create(ctx, obj); err != nil {
-		// The create call is idempotent, so if the object already exists
-		// then do not consider it to be an error.
-		if !apierrors.IsAlreadyExists(err) {
```

Comment:
I don't think we should drop this test, because doing so makes this func not re-entrant for the apply once strategy.
Also, if I'm not wrong, leaving this test should not be an issue for the new reconcile mode as well

Reply:
Since this comment seems to propose a similar path as the previous one, let's continue the conversation [here](https://github.com/kubernetes-sigs/cluster-api/pull/7497#discussion_r1064729651) and I will update both place accordingly.